File: python-examples/16-scheduled-workflow/worker.py
"""
Polos Worker for the Scheduled Workflow example.

Run with:
    python worker.py
"""

import asyncio
import os

from dotenv import load_dotenv
from polos import PolosClient, Worker

from workflows import (
    daily_cleanup,
    morning_report,
    hourly_sync,
    schedulable_reminder,
)

load_dotenv()


async def main():
    """Main function to run the worker."""
    project_id = os.getenv("POLOS_PROJECT_ID")
    if not project_id:
        raise ValueError(
            "POLOS_PROJECT_ID environment variable is required. "
            "Get it from the Polos UI at http://localhost:5173/projects/settings"
        )

    client = PolosClient(
        project_id=project_id,
        api_url=os.getenv("POLOS_API_URL", "http://localhost:8080"),
    )

    worker = Worker(
        client=client,
        workflows=[
            daily_cleanup,
            morning_report,
            hourly_sync,
            schedulable_reminder,
        ],
    )

    print("Starting Scheduled Workflow Examples worker...")
    print(f"  Project ID: {project_id}")
    print(f"  Workflows: {[w.id for w in worker.workflows]}")
    print("  Scheduled workflows:")
    for w in worker.workflows:
        if hasattr(w, "schedule") and w.schedule:
            schedule = w.schedule
            if isinstance(schedule, dict):
                cron = schedule.get("cron", "N/A")
                tz = schedule.get("timezone", "UTC")
                print(f"    - {w.id}: {cron} ({tz})")
            elif isinstance(schedule, str):
                print(f"    - {w.id}: {schedule} (UTC)")
            elif schedule is True:
                print(f"    - {w.id}: (schedulable dynamically)")
    print("  Press Ctrl+C to stop\n")

    try:
        await worker.run()
    except KeyboardInterrupt:
        print("\nShutting down worker...")
        await worker.shutdown()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript-examples/16-scheduled-workflow/worker.ts
/**
 * Polos Worker for the Scheduled Workflow example.
 *
 * Run with:
 *   npx tsx worker.ts
 *
 * Environment variables:
 *   POLOS_PROJECT_ID - Your project ID (required)
 *   POLOS_API_URL - Orchestrator URL (default: http://localhost:8080)
 *   POLOS_API_KEY - API key for authentication (optional for local development)
 *   POLOS_DEPLOYMENT_ID - Deployment ID (default: scheduled-workflow-examples)
 */

import 'dotenv/config';
import { Worker } from '@polos/sdk';
import {
  dailyCleanup,
  morningReport,
  hourlySync,
  schedulableReminder,
} from './workflows.js';

async function main() {
  const projectId = process.env['POLOS_PROJECT_ID'];
  if (!projectId) {
    throw new Error(
      'POLOS_PROJECT_ID environment variable is required. ' +
        'Set it to your project ID (e.g., export POLOS_PROJECT_ID=my-project). ' +
        'You can get this from the output printed by `polos-server start` or from the UI page at ' +
        "http://localhost:5173/projects/settings (the ID will be below the project name 'default')",
    );
  }

  const apiUrl = process.env['POLOS_API_URL'] ?? 'http://localhost:8080';
  const apiKey = process.env['POLOS_API_KEY'] ?? '';
  const deploymentId = process.env['POLOS_DEPLOYMENT_ID'] ?? 'scheduled-workflow-examples';

  const allWorkflows = [
    dailyCleanup,
    morningReport,
    hourlySync,
    schedulableReminder,
  ];

  const worker = new Worker({
    apiUrl,
    apiKey,
    projectId,
    deploymentId,
    workflows: allWorkflows,
  });

  console.log('Starting Scheduled Workflow Examples worker...');
  console.log(`  Project ID: ${projectId}`);
  console.log(`  Workflows: [${allWorkflows.map((w) => w.id).join(', ')}]`);
  console.log('  Scheduled workflows:');
  for (const w of allWorkflows) {
    const schedule = w.config.schedule;
    if (typeof schedule === 'string') {
      console.log(`    - ${w.id}: ${schedule} (UTC)`);
    } else if (schedule && typeof schedule === 'object') {
      console.log(`    - ${w.id}: ${schedule.cron} (${schedule.timezone ?? 'UTC'})`);
    } else {
      console.log(`    - ${w.id}: (schedulable dynamically)`);
    }
  }
  console.log('  Press Ctrl+C to stop\n');

  await worker.run();
}

main().catch(console.error);
